core/testcontainers/core/container.py

```diff
@@ -36,29 +46,32 @@ def __init__(self, image: str, docker_client_kw: Optional[dict] = None, **kwargs
         self._name = None
         self._kwargs = kwargs
 
-    def with_env(self, key: str, value: str) -> "DockerContainer":
+    def with_env(self, key: str, value: str) -> DockerContainer:
         self.env[key] = value
         return self
 
-    def with_bind_ports(self, container: int, host: Optional[int] = None) -> "DockerContainer":
+    def with_bind_ports(self, container: int, host: Optional[int] = None) -> DockerContainer:
         self.ports[container] = host
         return self
 
-    def with_exposed_ports(self, *ports: int) -> "DockerContainer":
+    def with_exposed_ports(self, *ports: int) -> DockerContainer:
         for port in ports:
             self.ports[port] = None
         return self
 
-    def with_kwargs(self, **kwargs) -> "DockerContainer":
+    def with_kwargs(self, **kwargs) -> DockerContainer:
         self._kwargs = kwargs
         return self
 
-    def maybe_emulate_amd64(self) -> "DockerContainer":
+    def maybe_emulate_amd64(self) -> DockerContainer:
         if is_arm():
             return self.with_kwargs(platform="linux/amd64")
         return self
 
-    def start(self) -> "DockerContainer":
+    def start(self) -> DockerContainer:
+        if not RYUK_DISABLED and self.image != RYUK_IMAGE:
+            logger.debug("Creating Ryuk container")
+            Reaper.get_instance()
```

Comment:
AFAIK in Python you don't need to tie singletons to classes. You would have guarantee it existing because even if you `from my_module import MyClass` then the contents of `my_module` would still be evaluated -> the singleton would exist.

ie. at the end of the file, you can:
```
reaper = Reaper()
```
and in the quoted section here `reaper.start()` or similar (on the same condition you set above)


However, it's good enough and the project isn't free of other things that look a bit like java-isms to me.
(also, happy to be proven wrong if you can point at other python projects that do it this way)

Reply:
i think its best to avoid breaking changes in principle - like to introduce something configurable and maybe print deprecations or warnings or something

Reply:
I don't thinks its a good practice to have side-effects on imports, even in this case where it would be to create an instance of `Reaper` that doesn't do much setup in its `__init__` method. It limits what can be done in the future regarding changes to initialization without breaking compatibility. Therefore I put it onto a class so that we have "namespaced" the functionality into a single collection of fields and methods, creating a nice API where you can import `Reaper` anywhere in your codebase and use the singleton as you want, without any side-effects.

Reply:
> I don't thinks its a good practice to have side-effects on imports

This is why I said in other threads that since your implementation works, we can leave it like so. What was and still is odd for me is the fact that I've not seen other modern projects out there that do this. That's all!

I'll close this conversation because we started mixing topics - the "is it good enough" is answered: yes, it is good enough.